feat: Adding support for i18n (closes #462)

Mara-Li opened the pull request:
Allow translating the website using configuration (new key "locale")

> [!note]
> The scope of the PR is making a translation of the **interface** of Quartz

I will obviously support all plugin that want to create i18n support for quartz, but it will out of the scope of this PR.

Mara-Li committed:
fix: alt error mix with height/width

More granular detection of alt and resize in image

Mara-Li committed:
Merge branch 'jackyzha0:v4' into v4

Mara-Li committed:
fix: format

Mara-Li committed:
feat: init i18n

Mara-Li committed:
feat: add translation

Mara-Li committed:
style: prettier for test

jackyzha0 commented:
Hey, thanks for this! I'll take a look when I have time

Mara-Li committed:
fix: build-up the locale to fusion with dateLocale

Mara-Li committed:
style: run prettier

aarnphm commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "GitHooks.hooksDirectory": "e:\\Documents\\Github\\quartz\\.git\\hooks"
+}
```
```suggestion
```

Let's remove this file.

aarnphm commented on quartz/components/Head.tsx:
```diff
@@ -1,11 +1,12 @@
   function Head({ cfg, fileData, externalResources }: QuartzComponentProps) {
-    const title = fileData.frontmatter?.title ?? "Untitled"
-    const description = fileData.description?.trim() ?? "No description provided"
+    const title = fileData.frontmatter?.title ?? i18next.t("head.untitled")
+    const description = fileData.description?.trim() ?? i18next.t("head.noDescriptionProvided")
```
should we also add options to parse i18n from frontmatter as well?

I imagine users writing their own file would be a lot nicer to do something like this

```markdown
---
title: chaos is a state of mind.
i18n:
  fr: le chaos est un état d'esprit
  es: el caos es un estado mental
---
```

Users can also set a default language in `quartz.config.ts` as well.

aarnphm replied:
edit: shoudl we make this into a plugin instead?. We can probably parse different h3 nodes and others part from a transformers standpoint as well, instead of manually doing this.

I wonder this would get out of hand pretty fast as well. But I don't have an opinion on this other than maintainability of quartz.

aarnphm replied:
probably a bit also out of scope, but would be nice to have a way to change locale as well 🤔

lucastucious replied:
I think we should split. If this PR is for  "quartz system" (interface), we could have another PR as a plugin to localize user content

aarnphm replied:
yes that makes sense. Feel free to ping me I can help with the review.

aarnphm reviewed (commented):
my two cents. also make sure to run `npm run format` as well 😄 

but definitely get jacky's opinion once he's available.

lucastucious commented:
I think we need a way to handle localized notes/pages. 

What if I want to write content and localize it in English.

And as it's a major feature, we should update docs in this PR.

But as @aarnphm say, it's probably out of scope for this PR

lucastucious commented on quartz/components/Graph.tsx:
```diff
@@ -2,6 +2,7 @@ import { QuartzComponentConstructor, QuartzComponentProps } from "./types"
 // @ts-ignore
 import script from "./scripts/graph.inline"
 import style from "./styles/graph.scss"
+import i18next from "i18next"
```
Maybe all QuartzComponentProps should import i18next

lucastucious commented on quartz/i18n/locales/fr.json:
```diff
@@ -1,9 +1,37 @@
+    "untitled": "Sans titre"
+  },
+  "recentNote": {
+    "seeRemainingMore": "Voir {{- remaining}} more"
```
Petit oubli de traduction 🫢.
"more" -> "plus"

jackyzha0 commented on package.json:
```diff
@@ -50,6 +50,7 @@
     "hast-util-to-html": "^9.0.0",
     "hast-util-to-jsx-runtime": "^2.3.0",
     "hast-util-to-string": "^3.0.0",
+    "i18next": "^23.7.20",
```
i actually think because we do i18n at compile time, we can get away with not using an extra library here at all and just simply do a lookup into that resources object that you created with some string interpolation

each file will probably be something like

```ts
const enTranslations = {
  seeRemaining: ({remaining}) => `See {remaining} more`
}
```

most i18n libraries are designed for client-side dynamic i18n so I think we can do better here

Mara-Li replied:
I use i18next because I think it make easy to maintains, specially under VSCode. More over, using crowdin, we could also support to translate in multiple language if we use JSON file instead of ts file for translation.

Maybe we could make something that use a local function to translate `{{variable}}` without using i18next ?

Mara-Li replied:
Also, I don't know how I can build the default main language without using i18next :<, as some component doesn't load config file

Mara-Li replied:
About the i18n function, maybe we could tweaks around types to allow at last autocompletion ?
The problem is : if we add new key, we must edit manually the two JSON + know the key set.

Using i18n prevent this problem as VSC already know and the plugin i18n works for that.

Mara-Li committed:
remove cursed file

Mara-Li committed:
refactor: remove i18n library and use locale way instead

Mara-Li committed:
format with prettier

Mara-Li committed:
forgot to remove test

Mara-Li committed:
prevent merging error

Mara-Li committed:
Merge remote-tracking branch 'upstream/v4' into i18n-support

Mara-Li committed:
format

Mara-Li committed:
format

Mara-Li committed:
fix: allow string for locale
- Check during translation if valid / existing locale
- Allow to use "en" and "en-US" for example
- Add fallback directly in the function
- Add default key in the function
- Add docstring to cfg.ts

Mara-Li committed:
forgot item translation

Mara-Li committed:
remove unused locale variable

Mara-Li committed:
Merge branch 'v4' into i18n-support

Mara-Li commented:
as locale is fusionned with the date formatting, I updated the i18n translation function to search for `en-US` first and fallback if the string is set to `en`
Allow, for example, to have different translation for country (fr-CA, zh-TW) and use fallback if the translation doesn't exist (yet).

aarnphm commented on quartz.config.ts:
```diff
@@ -9,6 +9,7 @@ const config: QuartzConfig = {
     analytics: {
       provider: "plausible",
     },
+    locale: "fr-FR",
```
```suggestion
    locale: "en-US",
```

let's set this by default.

Mara-Li replied:
Oh, yes, sorry, forgot to remove that!

Mara-Li committed:
forgot to remove fr-FR testing

Mara-Li committed:
Merge branch 'v4' into i18n-support

jackyzha0 commented:
hey, can you resolve the merge conflicts? i'll try to review tmrw or on thurs

aarnphm commented on quartz/i18n/i18next.ts:
```diff
@@ -0,0 +1,37 @@
+    ) ?? "en-US"
+  const getTranslation = (key: string) => {
+    const keys = key.split(".")
+    let translationString: string | Record<string, unknown> =
```
```suggestion
    let translationString: string | Record<string, Record<string, string>> =
```

🤔

Mara-Li committed:
Merge branch 'v4' into i18n-support

Mara-Li committed:
format

Mara-Li committed:
Merge branch 'v4' into i18n-support

Mara-Li committed:
Merge branch 'v4' into i18n-support

jackyzha0 commented:
Thanks for this, almost exactly what I was looking for! There's a few more minor nits but I'll address them myself after I merge
